Documentation/sources/how-to-messaging-delegate.md

```diff
@@ -0,0 +1,147 @@
+# Programmatically control the display of in-app messages
+
+You can now implement a `MessagingDelegate` in order to be alerted when specific events occur during the lifecycle of an in-app message.
+
+### Register the delegate with the Adobe Service Provider
+
+The `ServiceProvider` class maintains an optional property that holds reference to the `FullscreenMessaageDelegate`.
+
+```java
+// defined in public class ServiceProvider 
+public void setMessageDelegate(FullscreenMessageDelegate messageDelegate)
+```
+
+The `setMessageDelegate` API is then called with a custom `MessagingDelegate` object which must extend the `MessagingDelegate` class.
+
+```java
+CustomDelegate myMessagingDelegate = new CustomDelegate();
+ServiceProvider.getInstance().setMessageDelegate(myMessagingDelegate);
+```
+
+### FullscreenMessageDelegate interface
+
+The `FullscreenMessageDelegate` interface, which is implemented in the Android Messaging extension in the `MessagingDelegate` class, is defined below:
+
+```java
+/**
+ * Delegate for Messaging extension in-app message events.
+ */
+public interface FullscreenMessageDelegate {
+	/**
+	 * Invoked when the in-app message is displayed.
+	 *
+	 * @param message FullscreenMessage the in-app message being displayed
+	 */
+	void onShow(final FullscreenMessage message);
+
+	/**
+	 * Invoked when the in-app message is dismissed.
+	 *
+	 * @param message FullscreenMessage the in-app message being dismissed
+	 */
+	void onDismiss(final FullscreenMessage message);
+
+	/**
+	 * Used to determine if the in-app message should be shown.
+	 *
+	 * @param message FullscreenMessage the in-app message that is about to get displayed
+	 */
+	boolean shouldShowMessage(final FullscreenMessage message);
+
+	/**
+	 * Invoked when the in-app message is attempting to load a url.
+	 *
+	 * @param message FullscreenMessage the in-app message attempting to load the url
+	 * @param url     String the url being loaded by the message
+	 *
+	 * @return True if the core wants to handle the URL (and not the fullscreen message view implementation)
+	 */
+	boolean overrideUrlLoad(final FullscreenMessage message, final String url);
+
+	/**
+	 * Invoked when the in-app message failed to be displayed.
+	 */
+	void onShowFailure();
+}
+```
+
+### Retrieving the Message object from the implemented interface methods
+
+The user interface methods (except for `onShowFailure()`) in a `FullscreenMessageDelegate` implementation will be passed an `AEPMessage` object. An `AEPMessage` object is the Android Core implementation of the `FullscreenMessage` interface. It contains a reference to the parent `Message` class and is your primary way to interact with the message.
+
+A reference to the `AEPMessage` object can be obtained by calling `fullscreenMessage.getParent()` . An example of how to access the `Message` in the `onShow` delegate method can be seen below:
```

Comment:
are all these references to `AEPMessage` correct?

Reply:
good call out, i had to clean this up